Zend/zend_partial.c

```diff
@@ -0,0 +1,1129 @@
+					required = num;
+				}
+				memcpy(info,
+					&partial->func.common.arg_info[offset],
+					sizeof(zend_arg_info));
+				if (EXPECTED(num < MAX_ARG_FLAG_NUM)) {
+					uint32_t mode = ZEND_ARG_SEND_MODE(info);
+					if (mode) {
+						ZEND_SET_ARG_FLAG(&partial->trampoline, num, mode);
+						byref = true;
+					}
+				}
+				info++;
+			} else {
+				ZEND_ASSERT(ZEND_PARTIAL_FUNC_FLAG(&partial->func, ZEND_ACC_VARIADIC));
+				num++;
+				/* Placeholders that run into the variadic portion become
+				 * required and make all params before them required */
+				required = num;
+				info->name = zend_empty_string;
+				info->default_value = NULL;
+				if (ZEND_PARTIAL_IS_CALL_TRAMPOLINE(&partial->func)) {
+					info->type = (zend_type){0};
+				} else {
+					info->type = (partial->func.common.arg_info + partial->func.common.num_args)->type;
+					ZEND_TYPE_FULL_MASK(info->type) &= ~_ZEND_IS_VARIADIC_BIT;
+				}
+				if (EXPECTED(num < MAX_ARG_FLAG_NUM)) {
+					uint32_t mode = ZEND_ARG_SEND_MODE(info);
+					if (mode) {
+						ZEND_SET_ARG_FLAG(&partial->trampoline, num, mode);
+						byref = true;
+					}
+				}
+				info++;
+			}
+		} else if (Z_ISUNDEF_P(arg)) {
+			ZEND_ASSERT(!ZEND_PARTIAL_CALL_FLAG(partial, ZEND_APPLY_VARIADIC));
+			ZEND_ADD_CALL_FLAG(partial, ZEND_APPLY_UNDEF);
+		} else {
+			if (num > required) {
+				ZEND_ADD_CALL_FLAG(partial, ZEND_APPLY_UNDEF);
+			}
+		}
+		offset++;
+	}
+
+	/* If we have a variadic placeholder and the underlying function is
+	 * variadic, add a variadic param. Otherwise, allow extra args to be passed
+	 * (implied by the ZEND_APPLY_VARIADIC flag) but don't add an explicit
+	 * variadic param. */
+	if (ZEND_PARTIAL_CALL_FLAG(partial, ZEND_APPLY_VARIADIC)
+			&& ZEND_PARTIAL_FUNC_FLAG(&partial->func, ZEND_ACC_VARIADIC)) {
+		if (ZEND_PARTIAL_IS_CALL_TRAMPOLINE(&partial->func)) {
+			memcpy(info, zend_call_magic_arginfo, sizeof(zend_arg_info));
+		} else {
+			memcpy(info,
+					partial->func.common.arg_info + partial->func.common.num_args,
+					sizeof(zend_arg_info));
+		}
+		ZEND_PARTIAL_FUNC_ADD(&partial->trampoline, ZEND_ACC_VARIADIC);
+		uint32_t mode = ZEND_ARG_SEND_MODE(info);
+		if (mode) {
+			for (uint32_t i = num + 1; i < MAX_ARG_FLAG_NUM; i++) {
+				ZEND_SET_ARG_FLAG(&partial->trampoline, i, mode);
+			}
+			byref = true;
+		}
+		info++;
+	}
+
+	if (byref) {
+		ZEND_ADD_CALL_FLAG(partial, ZEND_APPLY_BYREF);
+	}
+
+	partial->trampoline.common.num_args = num;
+	partial->trampoline.common.required_num_args = required;
+	partial->trampoline.common.arg_info = signature;
+
+	if (ZEND_PARTIAL_FUNC_FLAG(&partial->trampoline, ZEND_ACC_HAS_RETURN_TYPE)) {
+		partial->trampoline.common.arg_info++;
+	}
+
+	if (partial->func.type == ZEND_INTERNAL_FUNCTION
+			&& !ZEND_PARTIAL_FUNC_FLAG(&partial->func, ZEND_ACC_USER_ARG_INFO)) {
+		/* Internal functions use an incompatible arg_info struct
+		 * TODO: unify zend_arg_info / zend_internal_arg_info? */
```

Comment:
This would be nice to do, but that would require allocating zend_strings rather than using C strings which are inside the binary. Would make a lot of other stuff simpler tho.